controllers/nstemplateset/namespaces.go

```diff
@@ -275,3 +287,97 @@ func getNamespaceName(request reconcile.Request) (string, error) {
 	}
 	return namespace, nil
 }
+
+// isUpToDateAndProvisioned checks if the obj has the correct Template Reference Label.
+// If so, it processes the tier template to get the expected roles and rolebindings and then checks if they are actually present in the namespace.
+func (r *namespacesManager) isUpToDateAndProvisioned(ns *corev1.Namespace, tierTemplate *tierTemplate) (bool, error) {
+	if ns.GetLabels() != nil &&
+		ns.GetLabels()[toolchainv1alpha1.TierLabelKey] == tierTemplate.tierName &&
+		ns.GetLabels()[toolchainv1alpha1.TemplateRefLabelKey] == tierTemplate.templateRef {
+
+		newObjs, err := tierTemplate.process(r.Scheme, ns.GetLabels()[toolchainv1alpha1.OwnerLabelKey], template.RetainAllButNamespaces)
```

Comment:
Agreed. Yeah i can create a new story to work on that. The caching I used did actually solve that. But maybe instead of having a placeholder user, I can cache with the user itself. The only question then would be the number of namespaces that are not up-to-date is more than the namespaces which are up-to-date, and the impact of caching all namespaces on the memory. I can look at that for sure.

Reply:
@MatousJobanek the original idea of the cache was to cache a small number of the processed templates so we would avoid processing the same template for different users over and over again when starting the operator (N times per template where N is the number of the users). But the problem with that is that the processed template is actually user specific (we pass the username as a parameter). There are some options to work it around for this specific scenario but that cache won't be useful in other use cases.
Creating a cache for the user-specific templates IMO would make the cache too big and won't help with the operator restarting at all. Most likely such cache would require more compute resources than the non-cache version.
And since the performance testing actually showed that there is no any resource usage between a non-cache and the cache one (some semi-fake cache was used just for the testing purposes) we decided to drop the cache thingy completely.

Reply:
Thank you for summarizing it so well Alexey.

Reply:
Thanks for the explanation - yeah, this is exactly what I expected that you did. 

My proposal wasn't to introduce an additional cache. The point was to store the list of the processed objects within this single reconcile loop - so it doesn't process the same template for the same user in the same reconcile twice.
But it's a tiny improvement and as was already mentioned before - the perf tests are fine, so we don't have to spend so much time with it  :-).

Reply:
Ah... I see what you mean now. Yes it does make sense for sure what you are talking about.